lib/Storage/MySqlTimeSeriesStore.php

```diff
@@ -0,0 +1,371 @@
+
+        // Run our query using a connection object (to save memory)
+        $connection = $this->store->getConnection();
+        $statement = $connection->prepare($sql);
+
+        // Execute
+        $statement->execute($params);
+
+        $rows = [];
+
+        while ($row = $statement->fetch(\PDO::FETCH_ASSOC)) {
+
+            $entry = [];
+            // Read the columns
+            $entry['type'] = $row['type'];
+            $entry['start'] = $row['start'];
+            $entry['end'] = $row['end'];
+            $entry['layout'] = $row['Layout'];
+            $entry['display'] = $row['Display'];
+            $entry['media'] = $row['MediaName'];
+            $entry['tag'] = $row['tag'];
+            $entry['displayId'] = $row['displayId'];
+            $entry['layoutId'] = $row['layoutId'];
+            $entry['widgetId'] = $row['widgetId'];
+            $entry['mediaId'] = $row['mediaId'];
+
+            $rows[] = $entry;
+
+        }
+
+        return ['statData'=> $rows];
+
+    }
+
+    /** @inheritdoc */
+    public function deleteStats($maxage, $fromDt = null, $options = [])
+    {
+        try {
+            $i = 0;
+            $rows = 1;
+
+            if($fromDt != null) {
+                $delete = $this->store->getConnection()
+                    ->prepare('DELETE FROM `stat` WHERE stat.statDate >= :fromDt AND stat.statDate < :toDt ORDER BY statId LIMIT :limit');
+
+                $delete->bindParam(':fromDt', $fromDt, \PDO::PARAM_STR);
+                $delete->bindParam(':toDt', $maxage, \PDO::PARAM_STR);
+                $delete->bindParam(':limit', $options['limit'], \PDO::PARAM_INT);
+            } else {
+                $delete = $this->store->getConnection()
+                    ->prepare('DELETE FROM `stat` WHERE stat.statDate < :maxage LIMIT :limit');
+                $delete->bindParam(':maxage', $maxage, \PDO::PARAM_STR);
+                $delete->bindParam(':limit', $options['limit'], \PDO::PARAM_INT);
+            }
+
+            while ($rows > 0) {
+
+                $i++;
+
+                // Run the delete
+                $delete->execute();
+
+                // Find out how many rows we've deleted
+                $rows = $delete->rowCount();
+
+                // We shouldn't be in a transaction, but commit anyway just in case
+                $this->store->commitIfNecessary();
+
+                // Give SQL time to recover
+                if ($rows > 0) {
+                    $this->log->debug('Stats delete effected ' . $rows . ' rows, sleeping.');
+                    sleep($options['statsDeleteSleep']);
+                }
+
+                // Break if we've exceeded the maximum attempts.
+                if ($i >= $options['maxAttempts'])
+                    break;
+            }
+
+            $this->log->debug('Deleted Stats back to ' . $maxage . ' in ' . $i . ' attempts');
+
+            $message = ' - ' . __('Done.');
+
+            return ['message' => $message];
```

Comment:
Passing a message back would certainly work, but is it a good design pattern?

Would it be more "PHP" if it returned a count of records deleted or an exception on failure (we'd have to throw a general exception rather than a PDO one as other time series stores might not be PDO).

We could then have our message set by the calling method as before.